ParseLoginUI/src/main/java/com/parse/ParseQueryRecyclerViewAdapter.java

```diff
@@ -0,0 +1,611 @@
+                }
+                if (!imageViewSet.containsKey(viewHolder.imageView)) {
+                    imageViewSet.put(viewHolder.imageView, null);
+                }
+                viewHolder.imageView.setPlaceholder(placeholder);
+                viewHolder.imageView.setParseFile((ParseFile) object.get(imageKey));
+                viewHolder.imageView.loadInBackground();
+            }
+            viewHolder.textView.setOnClickListener(new View.OnClickListener() {
+                @Override
+                public void onClick(View v) {
+                    if (onClickListener != null) {
+                        onClickListener.onClick(object, position);
+                    }
+                }
+            });
+        }
+        else if (position == objects.size()) {
+            viewHolder.textView.setText("Load more...");
+            if (viewHolder.imageView != null) {
+                viewHolder.imageView.setVisibility(View.GONE);
+            }
+            viewHolder.textView.setOnClickListener(new View.OnClickListener() {
+                @Override
+                public void onClick(View v) {
+                    loadNextPage();
+                }
+            });
+        }
+        else {
+            throw new RuntimeException();
+        }
+    }
+
+    @Override
+    public int getItemCount() {
+        int count = objects.size();
+
+        if (shouldShowPaginationCell()) {
+            count++;
+        }
+
+        return count;
+    }
+
+    @Override
+    public void registerAdapterDataObserver(RecyclerView.AdapterDataObserver observer) {
+        super.registerAdapterDataObserver(observer);
+        dataSetObservers.put(observer, null);
+        if (autoload) {
+            loadObjects();
+        }
+    }
+
+    @Override
+    public void unregisterAdapterDataObserver(RecyclerView.AdapterDataObserver observer) {
+        super.unregisterAdapterDataObserver(observer);
+        this.dataSetObservers.remove(observer);
+    }
+
+
+    /**
+     * This is a helper function to sync the objects with objectPages. This is only used with the
+     * CACHE_THEN_NETWORK option.
+     */
+    private void syncObjectsWithPages() {
+        objects.clear();
+        for (List<T> pageOfObjects : objectPages) {
+            objects.addAll(pageOfObjects);
+        }
+    }
+
+    private int getPaginationCellRow() {
+        return objects.size();
+    }
+
+    private boolean shouldShowPaginationCell() {
+        return paginationEnabled && objects.size() > 0 && hasNextPage;
+    }
+
+    private void notifyOnLoadingListeners() {
+        for (OnQueryLoadListener<T> listener : onQueryLoadListeners) {
+            listener.onLoading();
+        }
+    }
+
+    private void notifyOnLoadedListeners(List<T> objects, Exception e) {
+        for (OnQueryLoadListener<T> listener : onQueryLoadListeners) {
+            listener.onLoaded(objects, e);
+        }
+    }
```

Comment:
Do we foresee developers dynamically adding/removing listeners often? If so, we might want to iterate across a thread-safe copy instead to prevent errors.